smartsim/launcher/lsf/lsfLauncher.py

```diff
@@ -0,0 +1,189 @@
+from ...utils import get_logger
+from ..launcher import WLMLauncher
+from ..step import BsubBatchStep, JsrunStep
+from ..stepInfo import LSFStepInfo
+from .lsfCommands import bkill, bjobs
+from .lsfParser import parse_bjobs_jobid, parse_step_id_from_bjobs, parse_bsub
+
+logger = get_logger(__name__)
+
+
+class LSFLauncher(WLMLauncher):
+    """This class encapsulates the functionality needed
+    to launch jobs on systems that use LSF as a workload manager.
+
+    All WLM launchers are capable of launching managed and unmanaged
+    jobs. Managed jobs are queried through interaction with with WLM,
+    in this case LSF. Unmanaged jobs are held in the TaskManager
+    and are managed through references to their launching process ID
+    i.e. a psutil.Popen object
+    """
+
+    # init in WLMLauncher, launcher.py
+
+    def create_step(self, name, cwd, step_settings):
+        """Create a LSF job step
+
+        :param name: name of the entity to be launched
+        :type name: str
+        :param cwd: path to launch dir
+        :type cwd: str
+        :param step_settings: batch or run settings for entity
+        :type step_settings: BatchSettings | RunSettings
+        :raises SSUnsupportedError: if batch or run settings type isnt supported
+        :raises LauncherError: if step creation fails
+        :return: step instance
+        :rtype: Step
+        """
+        try:
+            if isinstance(step_settings, JsrunSettings):
+                step = JsrunStep(name, cwd, step_settings)
+                return step
+            if isinstance(step_settings, BsubBatchSettings):
+                step = BsubBatchStep(name, cwd, step_settings)
+                return step
+            raise TypeError(
+                f"RunSettings type {type(step_settings)} not supported by LSF"
+            )
+        except SSConfigError as e:
+            raise LauncherError("Job step creation failed: " + str(e)) from None
+
+    def run(self, step):
+        """Run a job step through PBSPro
+
+        :param step: a job step instance
+        :type step: Step
+        :raises LauncherError: if launch fails
+        :return: job step id if job is managed
+        :rtype: str
+        """
+        if not self.task_manager.actively_monitoring:
+            self.task_manager.start()
+
+        cmd_list = step.get_launch_cmd()
+        step_id = None
+        task_id = None
+        if isinstance(step, BsubBatchStep):
+            # wait for batch step to submit successfully
+            rc, out, err = self.task_manager.start_and_wait(cmd_list, step.cwd)
+            if rc != 0:
+                raise LauncherError(f"Bsub batch submission failed\n {out}\n {err}")
+            if out:
+                step_id = parse_bsub(out)
+                logger.debug(f"Gleaned batch job id: {step_id} for {step.name}")
+        else:
+            # jsrun doesn't direct output for us.
+            out, err = step.get_output_files()
```

Comment:
No, you're right. I kept it like this because it was more reliable than JSF's output when many JSF calls were running in bg. Now that we found we can avoid that, I'm fine trusting `stdio_stdout` and `stdio_stderr`.